### PR TITLE
Switch to storing annotations in a static field

### PR DESCRIPTION
**Pull request by mhutch**
Although it's kind of icky to store instance data on a static field, it is implemented in a robust manner that should prevent any surprises, and the details are hidden from authors of CLIs and authors of subsystems.

A subsystem reference is no longer needed when annotating the CliSymbol objects in the grammar, which makes construction of a pipleline much simpler.

The fluent grammar construction API now looks like this:

```csharp
new Option<bool>("--greeting").WithDescription("The greeting")
```

Eventually we will be able to use extension properties:

```csharp
new Option<bool>("--greeting") {
  Description = "The greeting"
}
```

Note that we still support the `IAnnotationProvider` model that allows advanced CLI authors to lazily or dynamically provide annotations.

I'm still trying to figure out ways to enforce type safety when using `DefaultValue`/`DefaultValueCalculation` with the underlying annotation storage methods `SetAnnotation`/`TryGetAnnotation`. However, the `SetDefaultValue` and `SetDefaultValueCalculation` specialized setters will at least ensure safety for CLI authors.

### CODE REVIEW

File: src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    // support more than absolutely necessary.
+    //
+    // Instead, we have a single static ConditionalWeakTable that maps each CliSymbol to an AnnotationStorage dictionary,
+    // which is lazily created and added to the ConditionalWeakTable a single time for each CliSymbol. The individual
+    // annotations are stored in the AnnotationStorage dictionary, which uses no locks, so is fast, but is not safe to be
+    // modified from multiple threads.
+    //
+    // This is fine, as we will have the following well-defined threading behavior: an annotated grammar and pipeline may
+    // only be constructed/modified from a single thread. Once the grammar/pipeline instance is fully constructed, it may
+    // be safely used from multiple threads.
```
Comment (KathleenDollard) [inline src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs:53]: We might want to add that once constructed, it is not safe to modify the grammar if it is in use.

File: src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.CommandLine.Subsystems.Annotations;
+
+/// <summary>
+/// Handles storage of annotations associated with <see cref="CliSymbol"/> instances.
+/// </summary>
+public static partial class AnnotationStorageExtensions
```
Comment (KathleenDollard) [inline src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs:12]: Does this class need to be public?

File: src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    /// <param name="symbol">The symbol that is annotated</param>
+    /// <param name="id">
+    /// The identifier for the annotation. For example, the annotation identifier for the help description is <see cref="HelpAnnotations.Description">.
+    /// </param>
+    /// <param name="provider">
+    /// An optional annotation provider that may implement custom or lazy construction of annotation values. Annotation returned by an annotation
+    /// provider take precedence over those stored in internal annotation storage.
+    /// </param>
+    /// <returns>The annotation value, if successful, otherwise <c>default</c></returns>
+    public static TValue? GetAnnotationOrDefault<TValue>(this CliSymbol symbol, AnnotationId<TValue> annotationId, IAnnotationProvider? provider = null)
```
Comment (KathleenDollard) [inline src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs:133]: I am a bit worried about this name as `default` has additional meaning in System.CommandLine

File: src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    {
+        if (symbol.TryGetAnnotation(annotationId, out TValue? value, provider))
+        {
+            return value;
+        }
+
+        return default;
+    }
+}
```
Comment (KathleenDollard) [inline src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs:142]: nit: NL

File: src/System.CommandLine.Subsystems/Subsystems/CliSubsystem.cs [hunk 5]
```diff
@@ -28,74 +29,26 @@ protected CliSubsystem(string name, SubsystemKind subsystemKind, IAnnotationProv
...
-    /// Attempt to retrieve the value for the symbol and annotation ID from the annotation provider.
+    /// Attempt to retrieve the <paramref name="symbol"/>'s value for the annotation <paramref name="id"/>. This will check the
+    /// annotation provider that was passed to the subsystem constructor, and the internal annotation storage.
     /// </summary>
     /// <typeparam name="TValue">The value of the type to retrieve</typeparam>
     /// <param name="symbol">The symbol the value is attached to</param>
-    /// <param name="id">The id for the value to be retrieved. For example, an annotation ID for help is description</param>
+    /// <param name="id">
+    /// The identifier for the annotation value to be retrieved.
+    /// For example, the annotation identifier for the help description is <see cref="HelpAnnotations.Description">.
```
Comment (KathleenDollard) [inline src/System.CommandLine.Subsystems/Subsystems/CliSubsystem.cs:42]: Is the type name `HelpAnnotation` used by the CLI author? IOW, do CLI authors ever have to type `HelpAnnotations.Description`? If they do, I think we should consider simplifying the name.

File: src/System.CommandLine.Subsystems/Subsystems/CliSubsystem.cs [hunk 6]
```diff
@@ -28,74 +29,26 @@ protected CliSubsystem(string name, SubsystemKind subsystemKind, IAnnotationProv
...
-    {
-        (_defaultProvider ??= new DefaultAnnotationProvider()).Set(symbol, id, value);
-    }
-
-    /// <inheritdoc cref="CliSubsystem.SetAnnotation{TValue}(CliSymbol, AnnotationId{TValue}, TValue)"/>
-    /// <param name="value">A delegate that returns the value.</param>
-    protected internal void SetAnnotation<TValue>(CliSymbol symbol, AnnotationId<Func<TValue>> id, Func<TValue> value)
-    {
-        (_defaultProvider ??= new DefaultAnnotationProvider()).Set<Func<TValue>>(symbol, id, value);
+        return symbol.TryGetAnnotation(id, out value, _annotationProvider);
```
Comment (KathleenDollard) [inline src/System.CommandLine.Subsystems/Subsystems/CliSubsystem.cs:51]: I'm not sure we still need this method. As a convenience, perhaps wait until we understand the calling code better.

**Review by KathleenDollard — APPROVED** [review]
This is great.

I have several comments for consideration, and some nits.

My vote is to merge this as soon as possible so I can accommodate it in some outstanding work (currently in a draft PR that needs rebasing and will be merge fun).

If you need me to re-approve after changes, ping me in teams or text so I can update your approval.

File: src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.AnnotationStorage.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.CommandLine.Subsystems.Annotations;
+
+partial class AnnotationStorageExtensions
```
[the remark this answers is not in the document]
Comment (mhutch) [reply src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.AnnotationStorage.cs:8]: It is a static class for extension methods. It's just a partial class so the nested class can be split into a separate file.

File: src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs (hunk 2 above)
Comment (mhutch) [reply src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs:12]: Yes. Implementors of subsystems need to use it to store any newly defined annotations. However, it's in the `System.CommandLine.Subsystems.Annotations` so it's not something most users will see.

File: src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs (hunk 3 above)
Comment (mhutch) [reply src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs:133]: Yes, but I'm not sure of a better name. This naming pattern is used by LINQ for `SingleorDefault` etc.

File: src/System.CommandLine.Subsystems/Subsystems/CliSubsystem.cs (hunk 5 above)
Comment (mhutch) [reply src/System.CommandLine.Subsystems/Subsystems/CliSubsystem.cs:42]: It's used `IAnnotationProvider` authors and subsystem authors. What would you simplify it to?

File: src/System.CommandLine.Subsystems/Subsystems/CliSubsystem.cs (hunk 6 above)
Comment (mhutch) [reply src/System.CommandLine.Subsystems/Subsystems/CliSubsystem.cs:51]: I will revise the comment. It's not just a convenience, it's the way that subsystem authors must access annotations if they want to respect the subsystem's `IAnnotationProvider`, as the provider is private.

File: src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.AnnotationStorage.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+{
+    class AnnotationStorage : IAnnotationProvider
+    {
+        record struct AnnotationKey(CliSymbol symbol, string annotationId);
```
[the remark this answers is not in the document]
Comment (mhutch) [reply src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.AnnotationStorage.cs:12]: Good catch, this was existing code (renamed from `DefaultAnnotationProvider`), but it turns out it was ignore the `Prefix` qualification. I restructured it to be clearer.

**Review by KathleenDollard — APPROVED** [review]

**Pull request merged**